test/test.py

```diff
@@ -0,0 +1,355 @@
+                with tempfile.TemporaryFile() as output_file:
+                    gfxd_input_buffer(data)
+
+                    gfxd_output_fd(output_file)
+
+                    gfxd_target(gfxd_f3dex2)
+                    gfxd_endian(GfxdEndian.big, 4)
+
+                    gfxd_execute()
+
+                    output_file.seek(0)
+                    self.assertEqual(expected, output_file.read().decode())
+
+    def test_gfxd_output_callback(self):
+        for sym, data, expected in self.data:
+            with self.subTest(sym):
+                output = io.BytesIO()
+
+                def callback(buf, count):
+                    self.assertEqual(count, len(buf))
+                    output.write(buf)
+                    return count
+
+                gfxd_input_buffer(data)
+
+                gfxd_output_callback(callback)
+
+                gfxd_target(gfxd_f3dex2)
+                gfxd_endian(GfxdEndian.big, 4)
+
+                gfxd_execute()
+
+                self.assertEqual(expected, output.getvalue().decode())
+
+
+class TestArgumentCallback(unittest.TestCase):
+    def setUp(self):
+        self.data: list[Sym, memoryview] = []
+        for sym in TEST_DATA.syms:
+            data = bytes(TEST_DATA.data[sym.offset :][: sym.size])
+            self.data.append((sym, data))
+
+    def test_gfxd_vtx_callback(self):
+        for sym, data in self.data:
+            expected = {
+                "emptyDList": [],
+                "oneTriDList": [(0x42042069, 3)],
+                "setLights1DList": [],
+            }[sym.name]
+            with self.subTest(sym):
+                gfxd_input_buffer(data)
+
+                gfxd_target(gfxd_f3dex2)
+                gfxd_endian(GfxdEndian.big, 4)
+
+                verts = []
+
+                def callback(vtx, num):
+                    verts.append((vtx, num))
+                    return 0
+
+                try:
+                    gfxd_vtx_callback(callback)
+
+                    gfxd_execute()
+                finally:
+                    def callback(vtx, num):
+                        return 0
+                    gfxd_vtx_callback(callback)
+
+                self.assertEqual(verts, expected)
+
+
+class TestMacroInfo(unittest.TestCase):
+    def setUp(self):
+        self.data: list[Sym, memoryview] = []
+        for sym in TEST_DATA.syms:
+            data = bytes(TEST_DATA.data[sym.offset :][: sym.size])
+            self.data.append((sym, data))
+
+    def test_gfxd_foreach_pkt(self):
+        for sym, data in self.data:
+            expected = {
+                "emptyDList": ["gsSPEndDisplayList"],
+                "oneTriDList": ["gsSPVertex", "gsSP1Triangle", "gsSPEndDisplayList"],
+                "setLights1DList": ["gsSPNumLights", "gsSPLight", "gsSPLight"],
+            }[sym.name]
+            with self.subTest(sym):
+                gfxd_input_buffer(data)
+
+                gfxd_target(gfxd_f3dex2)
+                gfxd_endian(GfxdEndian.big, 4)
```

Comment:
not sure about putting this everywhere :shrug: 
Probably deal with it if/when someone uses something other than the Superior :tm: Ucode f3dzex